### ISSUE TITLE
CMake/VS build can't find VS toolset starting on Feb 1 on Hosted VS2017 agents

### ISSUE THREAD

**Issue by cirvine-MSFT** [microsoft/azure-pipelines-image-generation#689]
**Describe the bug**
I have an Azure Pipeline build that uses CMake to generate and build VS 2017. This build started failing on Feb 1 with no changes to our code.
The error:
2019-02-06T20:44:34.2247861Z Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
2019-02-06T20:44:34.2248495Z Copyright (C) Microsoft Corporation. All rights reserved.
2019-02-06T20:44:34.2248798Z 
2019-02-06T20:44:34.2565590Z Build started 2/6/2019 8:44:34 PM.
2019-02-06T20:44:34.6346375Z Project "D:\a\1\s\bin\windows-x64-msvc-debug\ALL_BUILD.vcxproj" on node 1 (default targets).
2019-02-06T20:44:34.6346659Z Project "D:\a\1\s\bin\windows-x64-msvc-debug\ALL_BUILD.vcxproj" (1) is building "D:\a\1\s\bin\windows-x64-msvc-debug\ZERO_CHECK.vcxproj" (2) on node 1 (default targets).
2019-02-06T20:44:34.6354320Z C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Microsoft.Cpp.Platform.targets(65,5): error MSB8020: The build tools for Visual Studio 2010 (Platform Toolset = 'v100') cannot be found. To build using the v100 build tools, please install Visual Studio 2010 build tools.  Alternatively, you may upgrade to the current Visual Studio tools by selecting the Project menu or right-click the solution, and then selecting "Retarget solution". [D:\a\1\s\bin\windows-x64-msvc-debug\ZERO_CHECK.vcxproj]
2019-02-06T20:44:34.6372713Z Done Building Project "D:\a\1\s\bin\windows-x64-msvc-debug\ZERO_CHECK.vcxproj" (default targets) -- FAILED.
2019-02-06T20:44:34.7451860Z Done Building Project "D:\a\1\s\bin\windows-x64-msvc-debug\ALL_BUILD.vcxproj" (default targets) -- FAILED.
2019-02-06T20:44:34.7471947Z 
2019-02-06T20:44:34.7472165Z Build FAILED.
2019-02-06T20:44:34.7506236Z 
2019-02-06T20:44:34.7517338Z "D:\a\1\s\bin\windows-x64-msvc-debug\ALL_BUILD.vcxproj" (default target) (1) ->
2019-02-06T20:44:34.7517872Z "D:\a\1\s\bin\windows-x64-msvc-debug\ZERO_CHECK.vcxproj" (default target) (2) ->
2019-02-06T20:44:34.7518413Z (PlatformPrepareForBuild target) -> 
2019-02-06T20:44:34.7519993Z   C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Microsoft.Cpp.Platform.targets(65,5): error MSB8020: The build tools for Visual Studio 2010 (Platform Toolset = 'v100') cannot be found. To build using the v100 build tools, please install Visual Studio 2010 build tools.  Alternatively, you may upgrade to the current Visual Studio tools by selecting the Project menu or right-click the solution, and then selecting "Retarget solution". [D:\a\1\s\bin\windows-x64-msvc-debug\ZERO_CHECK.vcxproj]
2019-02-06T20:44:34.7520457Z 
2019-02-06T20:44:34.7520818Z     0 Warning(s)
2019-02-06T20:44:34.7521229Z     1 Error(s)
2019-02-06T20:44:34.7521576Z 
2019-02-06T20:44:34.7521927Z Time Elapsed 00:00:00.50
2019-02-06T20:44:34.8036225Z Running command failed: cmake --build D:\a\1\s\bin\windows-x64-msvc-debug --config debug

The toolset that should be targeted and found is v141

**Expected behavior**
VS Toolset v141 would be detected and used by visual studio.

**Actual behavior**
VS doesn't find any toolset and the build fails.

**Image impacted**
windows Hosted VS2017

**Comment by cirvine-MSFT** [comment]
I found a workaround that seems to have fixed the issue:
1) We are now calling vcvarsall.bat x64 before calling into CMake in our build script for both generate and build.
2) We needed to switch to using the architecture-specific VS CMake generator (Visual Studio 15 2017 Win64) instead of relying on ```set(CMAKE_GENERATOR_PLATFORM x64)```.

**Comment by degasus** [comment]
Might be related, I've likely triggered a similiar issue (v100 not found), but it was a bug on my side: I've missed to configure CMAKE_CONFIGURATION_TYPES or CMAKE_BUILD_TYPE. IMO still a valid bug as this kind of error message is terrible, but might you check if your issue is related?

**Comment by cirvine-MSFT** [comment]
@austin-jensen We were initializing `CMAKE_GENERATOR_PLATFORM`  in a toolchain file and that was working up until recently.

I have not confirmed if my workaround works without running vcvarsall.bat x64.

We have since moved to using Ninja as our generator/build system which has some other benifits as well. From my PoV this issue is resolved by moving to Ninja and feel free to close if there is no additional investigation necessary on your side. It is very likely we just happened to stumble upon a usage that was never intended to work, but happend to work for some time :)

@degasus I didn't get a chance to try your suggestion, but that is very good to know and thanks for adding the notes to this issue for others to find.
